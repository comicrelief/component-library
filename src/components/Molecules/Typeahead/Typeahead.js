import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { debounce } from 'lodash';

import TextInputWithDropdown from '../../Atoms/TextInputWithDropdown/TextInputWithDropdown';
import useStateObject from '../../../utils/useStateObject';

const initialState = {
  value: '',
  options: [],
  errorMsg: ''
};

// These just felt about right to me but could be changed.
const DELAY_DURATION = 200;
const MIN_CHARS_FOR_FETCH = 2;

const Typeahead = ({
  optionFetcher,
  optionParser,
  onSelect,
  id,
  label,
  name,
  dropdownInstruction,
  notFoundMessage,
  fetchErrorHandler,
  ...rest
}) => {
  const [state, updateState] = useStateObject(initialState);

  const debouncedFetch = useCallback(
    debounce(async value => {
      const valueTrimmed = value.trim();
      let { options, errorMsg } = initialState;
      if (valueTrimmed.length >= MIN_CHARS_FOR_FETCH) {
        try {
          options = await optionFetcher(valueTrimmed);
          if (options.length === 0) {
            errorMsg = notFoundMessage;
          }
        } catch (err) {
          errorMsg = fetchErrorHandler(err);
        }
        updateState({ options, errorMsg });
      }
    }, DELAY_DURATION),
    []
  );

  const onChange = e => {
    const { value } = e.currentTarget;
    // Resetting options / errorMsg as soon as the input changes seemed to me to be the nicest UX
    // (but happy to take advice on this!)
    const { options, errorMsg } = initialState;
    updateState({ value, options, errorMsg });
    debouncedFetch(value);
  };

  const { value, options, errorMsg } = state;

  return (
    <TextInputWithDropdown
      value={value}
      options={optionParser ? options.map(optionParser) : options}
      errorMsg={errorMsg}
      onChange={onChange}
      onSelect={(v, index) => {
        const selected = options[index];
        // pass the selected value up to the parent via callback
        onSelect(selected);
        // reset
        updateState(initialState);
      }}
      id={id}
      label={label}
      name={name}
      dropdownInstruction={dropdownInstruction}
      {...rest}
    />
  );
};

Typeahead.propTypes = {
  /** Takes the value and returns an array of options (e.g. by calling some lookup api.) */
  optionFetcher: PropTypes.func.isRequired,
  /**
   * Used to parse the options returned by the optionFetcher into strings - if they aren't already
   *  strings - which can then be displayed in the dropdown list.
   */
  optionParser: PropTypes.func,
  /** This function is used to provide data to the parent component when a selection is made. */
  onSelect: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  notFoundMessage: PropTypes.string.isRequired,
  /**
   * An optional error handling function which can either return a bespoke error message to be displayed here,
   *  or re-throw so the error can be handled further up the tree e.g. be an error boundary.
   */
  fetchErrorHandler: PropTypes.func,
  dropdownInstruction: PropTypes.string
};

Typeahead.defaultProps = {
  dropdownInstruction: null,
  optionParser: null,
  fetchErrorHandler: () =>
    'Sorry, there was an unexpected error. Please try again'
};

export default Typeahead;
