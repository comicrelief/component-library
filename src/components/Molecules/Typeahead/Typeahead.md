# Typeahead

```jsx
import axios from 'axios';
import Typeahead from './Typeahead';

const schoolsLookup = async query => {
  const response = await axios.get(
    'https://lookups.sls.comicrelief.com/schools/lookup',
    { params: { query } }
  );
  return response.data.data.schools.map(school => school.name);
};

<Typeahead
  optionFetcher={schoolsLookup}
  onSelect={school => alert(JSON.stringify(school, null, 2))}
  id="typeahead-test"
  label="Typeahead test"
  name="q"
  placeholder="Type to start searching..."
  notFoundMessage="Sorry, we couldn't find anything"
/>;
```
