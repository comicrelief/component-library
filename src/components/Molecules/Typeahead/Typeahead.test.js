import React from 'react';
import 'jest-styled-components';
import axios from 'axios';

import renderWithTheme from '../../../hoc/shallowWithTheme';
import Typeahead from './Typeahead';

const schoolsLookup = async query => {
  const response = await axios.get(
    'https://lookups.sls.comicrelief.com/schools/lookup',
    { params: { query } }
  );
  return response.data.data.schools.map(school => school.name);
};

it('renders correctly', () => {
  const renderer = renderWithTheme(
    <Typeahead
      optionFetcher={schoolsLookup}
      onSelect={() => {}}
      id="typeahead-test"
      label="Typeahead test"
      name="q"
      placeholder="Type to start searching..."
      notFoundMessage="Sorry, we couldn't find anything"
    />
  );

  expect(renderer.toJSON()).toMatchSnapshot();
});
