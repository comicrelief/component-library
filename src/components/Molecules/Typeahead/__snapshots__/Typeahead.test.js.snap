// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c2 {
  font-size: 1rem;
  line-height: 1rem;
  text-transform: inherit;
  font-weight: bold;
  line-height: normal;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: #5C5C5E;
}

.c3 {
  margin-bottom: 0.5rem;
}

.c5 {
  position: relative;
  box-sizing: border-box;
  width: 100%;
  height: 48px;
  padding: 1rem 1.5rem;
  background-color: #F4F3F5;
  border: 1px solid;
  border-color: #E1E2E3;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #000000;
  border-radius: 0.5rem;
  font-size: inherit;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c5:focus {
  border: 1px solid #666;
}

.c4 {
  position: relative;
  font-size: 1.25rem;
}

.c0 {
  position: relative;
}

@media (min-width:740px) {
  .c5 {
    max-width: 290px;
  }
}

<div
  className="c0 TextInputWithDropdown"
  onKeyDown={[Function]}
>
  <label
    className="c1 TextInputWithDropdown__input"
    htmlFor="typeahead-test"
  >
    <span
      className="c2 c3"
      color="inherit"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Typeahead test",
        }
      }
      size="s"
    />
    <div
      className="c4"
    >
      
      <input
        aria-describedby="typeahead-test"
        autoComplete="off"
        className="c5"
        id="typeahead-test"
        name="q"
        onChange={[Function]}
        placeholder="Type to start searching..."
        type="text"
        value=""
      />
      
    </div>
  </label>
</div>
`;
